Test fix of Location and LocationFactory.

DinoBektesevic opened the pull request:
If I run the following code

```
import lsst.daf.butler as dafButler

LocationFactory = dafButler.LocationFactory('datastoreRoot')

absloc = locationFactory.fromPath('/absolute/relative/file.ext')
relloc = locationFactory.fromPath('relative/file.ext')

print('path         ', absloc.path)
print('pathInStore  ', absloc.pathInStore)
print('uri          ', absloc.uri)
print('str          ', absloc)

print()
print('path         ', relloc.path)
print('pathInStore  ', relloc.pathInStore)
print('uri          ', relloc.uri)
print('str          ', relloc)

absloc = locationFactory.fromUri('file:///absolute/relative/file.ext')
relloc = locationFactory.fromUri('file://relative/file.ext')

print('\n\n')
print('path         ', absloc.path)
print('pathInStore  ', absloc.pathInStore)
print('uri          ', absloc.uri)
print('str          ', absloc)

print()
print('path         ', relloc.path)
print('pathInStore  ', relloc.pathInStore)
print('uri          ', relloc.uri)
print('str          ', relloc)
```

I get the following outputs
```
path          datastoreRoot/absolute/relative/file.ext
pathInStore   absolute/relative/file.ext
uri           file:///absolute/relative/file.ext
str           file:///absolute/relative/file.ext

path          datastoreRoot/relative/file.ext
pathInStore   relative/file.ext
uri           file:///relative/file.ext
str           file:///relative/file.ext
```
and
```
path          datastoreRoot/absolute/relative/file.ext
pathInStore   absolute/relative/file.ext
uri           file:///absolute/relative/file.ext
str           file:///absolute/relative/file.ext

path          datastoreRoot/file.ext
pathInStore   file.ext
uri           file://relative/file.ext
str           file://relative/file.ext
```

The problem is assuming that `file://relative/file.ext` is a valid URI. It seems to me from the behaviour of `urllib` that this is not correct. The following are correct `file://` URI's:
```
file://localhost/absolute/relative/file.ext
file:///absolute/relative/file.ext
```
The following URIs seem to be "incorrect"
```
file://relative/file.ext
file://file.ext`
```
This is the reason why `pathInStore` is returned differently depending on whether the `Location` was created by using `fromUri` or `fromPath`. The `relative` is interpreted as `localhost` instead.

This PR is a proposition of a solution to this that enables the `Location` to treat `file://relative` and `file:///absolute/relative...` paths differently by assuming that the `datastoreRoot` provided to the `LocationFactory` is either an absolute path or a path relative to current directory and that the `Location` path is then relative to that. I do not know if this is desired behaviour or not but it seems to follow the logic seen elsewhere in `daf_butler`. Naturally, checks for consistent root locations when dealing with absolute paths are now mandatory. 

Comments welcome.

DinoBektesevic committed:
Test fix of Location and LocationFactory.

DinoBektesevic committed:
Support for fragments added to Location.

I didn't ran tests and just assumed the original would work. It
didn't. Problem were paths and uri's ending in '#predicted' which
got scrubbed from the Location but shouldn't have. Added support
for uri fragments, which will now preserve the '#<fragment>' at
the end of paths.

urllib.parse.urljoin also kept scrubbing important bits from paths
so it was replaced with a string concatenation which should work as
long as people don't start sending uri's or otherwise non-POSIX
paths to `LocationFactory.fromPath` method.

timj commented:
I have a general comment on path manipulations in URLs. It seems like Python recommends that when you are manipulating the path from a URL you explicitly use `posixpath`.  From the docs:

>  Some of this can actually be useful on non-Posix systems too, e.g. for manipulation of the pathname component of URLs.

So for maximum portability we have to be careful and use `os.path` (and `os.sep`) when we are dealing with filenames in a local datastore, and `posixpath` (and `posixpath.sep`) when dealing with URLs.

DinoBektesevic committed:
Changed URI sensitive parts to explicitly use posixpath.

Changed the `Location.fromUri` path dealing functionality
to explicitly use `posixpath` module so that there are no misshaps
across platforms since URI paths are guaranteed to be POSIX
compliant. Splitting is also done on an explicitly given separator
instead of relying on `posixpath.sep` or `os.path.sep`. The
remaining local path related functionality is left OS dependant.

Trimmed some unnecessary comments.

DinoBektesevic commented:
I changed all URI path manipulation to explicitly use `posixpath` module. @ktlim told me about this but I forgot. 

I should also note that this then requires me to fix the `S3Location` in PR https://github.com/lsst/daf_butler/pull/159 to only ever explicitly use POSIX paths as it could be confusing if `S3Location.pathInStore` returns a Windows-style path but `S3Location.uri` the correct `s3://` uri. I don't believe that this should be done here as `Location` might refer to a local path that would then be OS dependent? I am known to make mistakes, however.

EDIT: I ran ahead of myself again. `Location.fromPath`, as is, won't change `/` for `\` when it receives a different style path. Should I add a `replace` there to make sure the path uses `\` in there for good measure?

timj commented:
I've looked around the [web a bit](https://tools.ietf.org/html/rfc3986#section-3.3) and relative URIs for files are meant to be written as `file:relative/path.ext`:

```python
>>> urllib.parse.urlparse("file:relative/path.ext")
ParseResult(scheme='file', netloc='', path='relative/path.ext', params='', query='', fragment='')
```
The problem is that when urllib puts that back together it doesn't round trip:
```python
>>> urllib.parse.urlparse("file:relative/path.ext").geturl()
'file:///relative/path.ext'
>>> urllib.parse.urlparse(urllib.parse.urlparse("file:///relative/path.ext").geturl())
ParseResult(scheme='file', netloc='', path='/relative/path.ext', params='', query='', fragment='')
```
indicating that `urllib` is not fit for purpose. I wonder if we are meant to use a different url manipulation library.

timj commented:
I've looked at the furl package and the uri package and they both fail to handle relative path URIs. I've filed issues as marrow/uri#11 and gruns/furl#115

timj commented:
@DinoBektesevic Reading this PR and #159 I think I need to play with this branch a bit. I worry that there are lots of hacks hidden around the place (and in #159) stemming from less than consistent thinking inside `Location` and `LocationFactory` (in particular the completely wrong line at the end using `file://` with a relative path).

timj commented:
I am closing this PR in favor of #167
